addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java

```diff
@@ -59,51 +65,132 @@ public void deactivate() {
     @Override
     protected void startScan() {
         logger.debug("Starting Homematic discovery scan");
+        enableInstallMode();
         loadDevices();
     }
 
+    /**
+     * Will set controller in <i>installMode==true</i>, but only if the bridge
+     * is ONLINE (e.g. not during INITIALIZATION).
+     */
+    private void enableInstallMode() {
+        try {
+            HomematicGateway gateway = bridgeHandler.getGateway();
+            ThingStatus bridgeStatus = null;
+            
+            if (bridgeHandler.getThing() != null) {
+                Thing bridge = bridgeHandler.getThing();
+                bridgeStatus = bridge.getStatus();
+                updateInstallModeDuration(bridge);
+            }
+            if (ThingStatus.ONLINE == bridgeStatus) {
+                gateway.setInstallMode(true, installModeDuration);
+
+                int remaining = gateway.getInstallMode();
+                if (remaining > 0) {
+                    setIsInInstallMode();
+                    logger.debug("Successfully put controller in install mode. Remaining time: {} seconds", remaining);
+                } else {
+                    logger.error("Controller did not accept requested install mode");
+                }
+            } else {
+                logger.info("Will not attempt to set controller in install mode, because bridge is not ONLINE.");
+            }
+        } catch (Exception ex) {
+            logger.error("Failed to set Homematic controller in install mode", ex);
+        }
+    }
+    
+    private void updateInstallModeDuration(Thing bridge) {
+        HomematicConfig config = bridge.getConfiguration().as(HomematicConfig.class);
+        installModeDuration = config.getInstallModeDuration();
+    }
+    
+    @Override
+    public int getScanTimeout() {
+        return installModeDuration;
+    }
+
     @Override
     public synchronized void stopScan() {
         logger.debug("Stopping Homematic discovery scan");
         if (bridgeHandler != null && bridgeHandler.getGateway() != null) {
+            disableInstallMode();
             bridgeHandler.getGateway().cancelLoadAllDeviceMetadata();
         }
         waitForScanFinishing();
         super.stopScan();
     }
+    
+    private void disableInstallMode() {
+        try {
+            synchronized (installModeSync) {
+                if (isInInstallMode) {
+                    isInInstallMode = false;
+                    installModeSync.notify();
+                    bridgeHandler.getGateway().setInstallMode(false, 0);
+                }
+            }
+            
+        } catch (Exception ex) {
+            logger.error("Failed to disable Homematic controller's install mode", ex);
```

Comment:
Just to be sure: We can only end up here if there is something seriously wrong (i.e. it can be considered a software bug) and it is a non-recoverable situation, correct?

Reply:
and here